glean-core/src/upload/mod.rs

```diff
@@ -276,6 +279,55 @@ impl PingUploadManager {
         }
     }
 
+    /// Enqueue pings that might have been cached.
```

Comment:
The bulk of the logic that enforces the quota is in this function. All the rest is to support that we can do what we do here.

I have some questions regarding this approach, though. I would like feedback before proceeding and adding the metric recording to this.

1. I wonder if it is the best idea to only check for the pending pings size when scanning the directory, or if I should check everytime a ping is stored. I remember asking this before, but I don't remember the answer. Looking at this implementation it really makes me question how effective this limitation is if we are not checking everytime a ping is stored. Doing that would require many changes, specially since a new `PingMaker` is created everytime we want to store a new ping and it holds no state. We would basically have to scan the pending pings folder everytime we store a new ping, which is suboptimal to say the least, or instantiate the PingMaker only once, accumulate the size of every stored ping and decrement the size of every deleted ping. Writing this down, I see that it is probably not that best way. I am sure there are other solutions if we decide to go this route.

2. I decided to not even check the size of the deletion request folder. I explain why in this docstring. I wonder if it would be best to sum the size of the pending_ping with the deletion_request directory instead and then enforce that the sum of both directories size doesn't exceed the quota.

Reply:
One more thing I just realized: this approch will not take into account the size of pings submitted after initialization of Glean (specifically the initialization of the upload manager inside Glean) when calculating the size of the folder.

Reply:
To 1:
I had to check the code again, but Firefox Telemetry on Desktop enforces that quota only on start (where it runs asynchronously and gets canceled on shutdown if not finished)

I think that's still a reasonable choice.
We should _assume_ that we 1) generate only a reasonable amount of pings while running, 2) most of the time are able to sent them out, 3) we already check that pings are <1mb (correct me if I'm wrong, we do this on storage, not on upload, right?)
And as you figured out doing it differently is much more complex.

Desktop tries to get the size to 90% of the quota if it was larger. It does so by deleting the oldest pings first.
Is that what you intend to do here as well?

To 2:
Sounds reasonable. There shouldn't even be multiple deletion-request pings around
(the only way this can happen if some user's device is offline and they keep toggling data upload, right?)

Reply:
Yes, I mostly tried to copy over what desktop does. It just looks very different because implementation in general here is very different.

> Desktop tries to get the size to 90% of the quota if it was larger. It does so by deleting the oldest pings first.
Is that what you intend to do here as well?

Yes, since the cached pings list is ordered by date, starting with the newest and ending with the oldest we will delete older pings instead of newer. I do not check if we are at 90% though, I guess that was an oversight when copying the logic over. Do you think that it makes sense to do that for glean too?

Reply:
One more point I forgot to raise, for telemetry we have this 15MB quota, but that is for desktop and we havea different quota for mobile. Should we also do that for glean? (e.g. by exposing the quota as a config option for Glean).

That would probably be a follow up. In any case, do you think a quota of 15MB as a starting point is good enough? Should we file a follow up to make the quota configurable?

Reply:
This discussion continued in: https://github.com/mozilla/glean/pull/1110#discussion_r465010205